File: engine/api/pipeline.py
"""Investigation pipeline — Strategy → Orchestrator → Reasoner → Validator → Calibrator.

Spec: docs/Specs.md §10.2, docs/04-InvestigationEngine.md §1-§9.

This module contains the full pipeline as a single async function with
explicit dependency injection.  The FastAPI endpoint in ``main.py`` calls
``run_investigation()`` with real deps; tests inject mocks.
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

import structlog

from api.schemas import (
    ConfidenceBreakdown,
    EvidenceRow,
    TimelineStep,
    Verdict,
)
from llm.prompts.reasoner import (
    Reasoner,
    ReasonerOutput,
    ReasonerResult,
    build_corrective_messages,
    build_messages,
)
from llm.validation import validate_citations
from orchestrator.calibrator import (
    CalibrationInputs,
    calibrate,
    compute_evidence_convergence,
)
from orchestrator.strategy import StrategyInputs, select_strategy
from orchestrator.tools import EvidenceAccumulator, ToolContext

if TYPE_CHECKING:
    from api.schemas import InvestigateRequest
    from llm.client import LLMClient, Message
    from orchestrator.loop import Orchestrator

logger = structlog.get_logger(__name__)


# === Tool verb map (Specs §11.5, Glossary §6) ============================

_TOOL_VERBS: dict[str, str] = {
    "policy_match": "Matched against rules",
    "report_velocity": "Checked report velocity",
    "user_history": "Pulled author history",
    "prior_actions": "Reviewed prior mod actions",
    "thread_context": "Read thread context",
}


# === Pipeline =============================================================


async def run_investigation(  # noqa: PLR0913
    *,
    req: InvestigateRequest,
    orchestrator: Orchestrator,
    llm: LLMClient,
    # Subreddit context (fetched by caller from DB or defaults).
    personality: str,
    region: str,
    rules: str,
    cold_start: bool,
    user_risk_tier: str,
    # Precomputed cheap signals for strategy selection.
    velocity_zscore: float,
    rule_match_score: float,
    tier_override: str,
) -> PipelineResult:
    """Run the full investigation pipeline.  Pure business logic — no I/O
    beyond what the orchestrator tools and LLM client perform.

    Returns a ``PipelineResult`` containing the ``Verdict`` plus metadata
    needed for persistence.
    """
    t0 = time.perf_counter()
    log = logger.bind(
        correlation_id=req.correlation_id,
        subreddit_id=req.subreddit_id,
    )

    # 1. Strategy selection
    strategy_inputs = StrategyInputs(
        reporter_count=req.report.reporter_count,
        velocity_zscore=velocity_zscore,
        user_risk_tier=user_risk_tier,  # type: ignore[arg-type]
        rule_match_score=rule_match_score,
        personality=personality,  # type: ignore[arg-type]
        tier_override=tier_override,  # type: ignore[arg-type]
        cold_start=cold_start,
    )
    decision = select_strategy(strategy_inputs)
    log.info("pipeline.strategy", tier=decision.tier, rationale=decision.rationale)

    # 2. Orchestrator — run tools
    context = ToolContext(
        subreddit_id=req.subreddit_id,
        correlation_id=req.correlation_id,
        target_kind=req.target.kind,
        target_id=req.target.id,
        target_body=req.target.body,
        target_author_id=req.target.author,
        reporter_count=req.report.reporter_count,
        rule_match_score=rule_match_score,
    )
    orch_result = await orchestrator.run(decision=decision, context=context)
    accumulator = orch_result.accumulator

    # 3. Reasoner — LLM call
    is_partial = orch_result.early_stopped and orch_result.stop_reason != "converged"
    messages = build_messages(
        accumulator=accumulator,
        personality=personality,
        region=region,
        rules=rules,
        target_kind=req.target.kind,
        target_id=req.target.id,
        reporter_count=req.report.reporter_count,
        tier=decision.tier,
        tools_run=orch_result.tools_run,
        is_partial=is_partial,
        cold_start=cold_start,
    )

    reasoner_result: ReasonerResult | None = None
    validation_flag = False

    if decision.reasoner_required:
        reasoner_result = await _reason_with_retry(
            llm=llm,
            messages=messages,
            accumulator=accumulator,
            correlation_id=req.correlation_id,
            log=log,
        )
        validation_flag = reasoner_result is None

    # 4. Extract signals for calibration
    reasoner_output = reasoner_result.output if reasoner_result else _fallback_output()
    rule_match_strength = _extract_rule_match_strength(accumulator)
    evidence_signals = _extract_evidence_signals(accumulator)

    # 5. Calibrate
    cal_inputs = CalibrationInputs(
        llm_self_report=reasoner_output.raw_confidence,
        evidence_convergence=compute_evidence_convergence(evidence_signals),
        subreddit_accuracy=0.5,  # TODO(I-3.4): compute from feedback history
        rule_match_strength=rule_match_strength,
        validation_passed=not validation_flag,
        cold_start=cold_start,
        is_partial=is_partial,
    )
    cal_result = calibrate(cal_inputs)

    # 6. Assemble verdict
    total_ms = int((time.perf_counter() - t0) * 1000)
    model_reasoner = reasoner_result.model if reasoner_result else ""
    input_tokens = reasoner_result.input_tokens if reasoner_result else 0
    output_tokens = reasoner_result.output_tokens if reasoner_result else 0
    cost_usd = reasoner_result.cost_usd if reasoner_result else 0.0

    timeline = _build_timeline(accumulator)
    top_evidence = _build_top_evidence(
        accumulator, reasoner_output.top_evidence_ids
    )

    verdict = Verdict(
        correlation_id=req.correlation_id,
        tier=decision.tier,
        risk_tier=reasoner_output.risk_tier,
        recommendation=reasoner_output.recommendation,
        calibrated_confidence=cal_result.calibrated_confidence,
        rationale=reasoner_output.rationale,
        top_evidence=top_evidence,
        timeline=timeline,
        confidence_breakdown=ConfidenceBreakdown(
            llm_self_report=cal_result.llm_self_report,
            evidence_convergence=cal_result.evidence_convergence,
            subreddit_accuracy=cal_result.subreddit_accuracy,
            rule_match_strength=cal_result.rule_match_strength,
        ),
        model_reasoner=model_reasoner,
        model_summarizer="",  # summarizer not used in current pipeline
        cost_usd=round(cost_usd, 6),
        latency_ms=total_ms,
        validation_flag=validation_flag,
        degraded=reasoner_result is None,
        cold_start=cold_start,
    )

    log.info(
        "pipeline.completed",
        tier=decision.tier,
        recommendation=reasoner_output.recommendation,
        calibrated_confidence=cal_result.calibrated_confidence,
        confidence_tier=cal_result.tier,
        latency_ms=total_ms,
        tools_run=orch_result.tools_run,
        validation_flag=validation_flag,
    )

    return PipelineResult(
        verdict=verdict,
        accumulator=accumulator,
        tier=decision.tier,
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        cost_usd=cost_usd,
        model_reasoner=model_reasoner,
        validation_flag=validation_flag,
        cold_start=cold_start,
    )


# === Reasoner with retry ==================================================


async def _reason_with_retry(
    *,
    llm: LLMClient,
    messages: list[Message],
    accumulator: EvidenceAccumulator,
    correlation_id: str,
    log: structlog.stdlib.BoundLogger,
) -> ReasonerResult | None:
    """Call the Reasoner, validate, retry once on failure.

    Returns the valid ``ReasonerResult``, or ``None`` if both attempts fail.
    """
    reasoner = Reasoner(llm)

    # First attempt
    try:
        result = await reasoner.reason(
            messages=messages, correlation_id=correlation_id
        )
    except Exception:
        log.warning("pipeline.reasoner.failed", attempt=1)
        return None

    validation = validate_citations(
        result.output.rationale,
        accumulator,
        cited_evidence_ids=result.output.cited_evidence_ids,
    )
    if validation.passed:
        return result

    log.warning(
        "pipeline.validation.failed",
        attempt=1,
        reason=validation.reason,
        details=validation.details,
    )

    # Corrective retry
    corrective = build_corrective_messages(
        prior_messages=messages,
        prior_response=result.raw_text,
        validation_reason=validation.reason,
        validation_details=str(validation.details),
    )
    try:
        result2 = await reasoner.reason(
            messages=corrective, correlation_id=f"{correlation_id}:retry"
        )
    except Exception:
        log.warning("pipeline.reasoner.failed", attempt=2)
        return None

    validation2 = validate_citations(
        result2.output.rationale,
        accumulator,
        cited_evidence_ids=result2.output.cited_evidence_ids,
    )
    if validation2.passed:
        return result2

    log.warning(
        "pipeline.validation.failed",
        attempt=2,
        reason=validation2.reason,
        details=validation2.details,
    )
    return None


# === Helpers ==============================================================


def _fallback_output() -> ReasonerOutput:
    """Degraded verdict when the Reasoner fails twice."""
    return ReasonerOutput(
        risk_tier="LOW",
        recommendation="NO_RECOMMENDATION",
        rationale=(
            "ModPilot was unable to produce a recommendation for this report. "
            "The evidence has been collected and is available for review [ev-1]."
        ),
        top_evidence_ids=["ev-1"],
        raw_confidence=0.0,
        cited_evidence_ids=["ev-1"],
        flags=["reasoner_failed"],
    )


def _extract_rule_match_strength(accumulator: EvidenceAccumulator) -> float:
    """Pull the max similarity from policy_match evidence, or 0.0."""
    for entry in accumulator.successful_entries():
        if entry.tool == "policy_match":
            matches = entry.detail.get("matches")
            if isinstance(matches, list) and matches:
                first = matches[0]
                if isinstance(first, dict) and "similarity" in first:
                    return float(first["similarity"])
    return 0.0


def _extract_evidence_signals(accumulator: EvidenceAccumulator) -> list[float]:
    """Build a list of tool-level signal strengths for convergence scoring."""
    signals: list[float] = []
    for entry in accumulator.successful_entries():
        detail = entry.detail
        # policy_match: use top similarity
        if entry.tool == "policy_match":
            matches = detail.get("matches")
            if isinstance(matches, list) and matches:
                first = matches[0]
                if isinstance(first, dict):
                    signals.append(float(first.get("similarity", 0.0)))
            else:
                signals.append(0.0)
        # report_velocity: normalize z_score (cap at 1.0)
        elif entry.tool == "report_velocity":
            raw_z = detail.get("z_score", 0.0)
            z = float(str(raw_z))
            signals.append(min(abs(z) / 5.0, 1.0))  # z=5 → 1.0
        else:
            # Generic: 0.5 for any successful tool without a numeric signal
            signals.append(0.5)
    return signals


def _build_timeline(accumulator: EvidenceAccumulator) -> list[TimelineStep]:
    """Convert Evidence Accumulator entries to timeline steps."""
    steps: list[TimelineStep] = []
    for entry in accumulator.entries():
        steps.append(
            TimelineStep(
                tool=entry.tool,
                verb=_TOOL_VERBS.get(entry.tool, f"Ran {entry.tool}"),
                status=entry.status,
                latency_ms=entry.latency_ms,
                evidence_ids=[entry.id],
            )
        )
    return steps


def _build_top_evidence(
    accumulator: EvidenceAccumulator,
    top_ids: list[str],
) -> list[EvidenceRow]:
    """Select top evidence rows by the Reasoner's top_evidence_ids."""
    rows: list[EvidenceRow] = []
    for ev_id in top_ids:
        entry = accumulator.by_id(ev_id)
        if entry is not None:
            rows.append(
                EvidenceRow(
                    id=entry.id,
                    summary=entry.summary,
                    tool=entry.tool,
                )
            )
    return rows[:3]


# === Result type ==========================================================


@dataclass(frozen=True)
class PipelineResult:
    """Everything the API handler needs to build the response + persist."""

    verdict: Verdict
    accumulator: EvidenceAccumulator
    tier: str
    input_tokens: int = 0
    output_tokens: int = 0
    cost_usd: float = 0.0
    model_reasoner: str = ""
    validation_flag: bool = False
    cold_start: bool = False

File: devvit-app/src/engine/pipeline.ts
// Investigation pipeline — Strategy → Orchestrator → Reasoner → Validator → Calibrator.
// Mirrors engine/api/pipeline.py:run_investigation, ported to in-process Devvit.

import { EvidenceAccumulator, ToolRegistry } from './accumulator';
import { calibrate, computeEvidenceConvergence } from './calibrator';
import { GeminiClient } from './llm/gemini';
import {
  Reasoner,
  buildCorrectiveMessages,
  buildMessages,
  type ReasonerOutput,
  type ReasonerResult,
} from './llm/reasoner';
import { validateCitations } from './llm/validator';
import { Orchestrator } from './loop';
import { getPreset } from './personalities';
import { selectStrategy, type StrategyInputs } from './strategy';
import { ensureSubredditProfile } from './store/subreddit';
import { getThreadMemory } from './store/threadMemory';
import { getUserMemory } from './store/userMemory';
import { persistInvestigation } from './store/investigation';
import { bumpInvestigationStats, readAlignmentSnapshot } from './store/stats';
import { computeRuleMatch } from './ruleMatch';
import { computePriority, priorityHeadline } from './priority';
import {
  buildRuleMatchDisplay,
  deriveAuthorSignal,
  deriveConfidenceFactors,
  deriveKeyFactors,
} from './explainability';
import { deriveEscalation } from './escalation';
import { ReportVelocityTool } from './tools/reportVelocity';
import { UserHistoryTool } from './tools/userHistory';
import { PriorActionsTool } from './tools/priorActions';
import { ThreadContextTool } from './tools/threadContext';
import type {
  Recommendation,
  RiskTier,
  ToolContext,
  ToolName,
  Verdict,
} from './types';

export type InvestigateInput = {
  correlationId: string;
  subredditId: string;
  target: {
    kind: 'comment' | 'post';
    id: string;
    // For posts: the headline string. Distinct from body — the title is
    // where character attacks like "Rohit is a clown" usually live, and
    // the Reasoner must see it to judge the content correctly. Comments
    // have no separate title.
    title: string;
    body: string;
    author: string;
  };
  reporterCount: number;
  threadId?: string;
  threadExcerpts?: string[];
};

const TOOL_VERBS: Record<ToolName, string> = {
  policy_match: 'Matched against rules',
  report_velocity: 'Checked report velocity',
  user_history: 'Pulled author history',
  prior_actions: 'Reviewed prior mod actions',
  thread_context: 'Read thread context',
};

export async function runInvestigation(args: {
  input: InvestigateInput;
  geminiApiKey: string;
}): Promise<Verdict> {
  const t0 = Date.now();
  const llm = new GeminiClient(args.geminiApiKey);

  // Registry — tools that need the LLM get it at construction.
  const registry = new ToolRegistry();
  registry.register(new ReportVelocityTool());
  registry.register(new UserHistoryTool());
  registry.register(new PriorActionsTool());
  registry.register(new ThreadContextTool(llm));

  // Guard: subreddit_id is load-bearing — every persisted key includes it.
  // Falling back to a literal string would silently route reads to the wrong
  // hash and yield "no rules configured" even after a successful save.
  if (!args.input.subredditId || !args.input.subredditId.startsWith('t5_')) {
    throw new Error(
      `runInvestigation: missing or malformed subreddit_id (${args.input.subredditId}). ` +
        'Caller must pass context.subredditId.',
    );
  }

  // Pull subreddit + user + thread state.
  const [profile, userMem, threadMem] = await Promise.all([
    ensureSubredditProfile(args.input.subredditId),
    getUserMemory(args.input.subredditId, args.input.target.author),
    args.input.threadId
      ? getThreadMemory(args.input.subredditId, args.input.threadId)
      : Promise.resolve(null),
  ]);

  console.log('engine.profile_loaded', {
    correlation_id: args.input.correlationId,
    sub_id: args.input.subredditId,
    personality: profile.personality,
    region: profile.region,
    rules_chars: profile.rules.length,
    rules_preview: profile.rules.slice(0, 120),
    tier_override: profile.tierOverride,
    cold_start_count: profile.coldStartCount,
  });
  // Lowered cold-start threshold — original 50 is fine for a long-lived sub
  // but too punitive for fresh demos. After 10 investigations the sub has
  // enough history that demoting confidence by 0.85 stops being honest.
  const coldStart = profile.coldStartCount < 10;
  const userRiskTier = userMem?.riskTier ?? 'new';
  const threadEscalated = threadMem
    ? threadMem.modActionsTaken > 0 || threadMem.escalationTurn !== null
    : false;

  // For posts, the title often carries the actual content judgment ("Rohit
  // is a clown" lives in the title; the body might be milder). The Reasoner
  // sees both; the rule-match precheck scans both. Comments have no title.
  const titlePlusBody = [args.input.target.title, args.input.target.body]
    .filter(Boolean)
    .join('\n')
    .trim();

  // 0. Rule-match precheck. Substring score over content words across configured
  //    rules. Feeds Strategy Selector + Calibrator and is appended as a real
  //    evidence row so the Reasoner can cite it.
  const ruleMatch = computeRuleMatch(titlePlusBody, profile.rules);
  console.log('engine.rule_match', {
    correlation_id: args.input.correlationId,
    score: ruleMatch.score,
    matched_rule: ruleMatch.matchedRule,
    matched_terms: ruleMatch.matchedTerms,
  });

  // 1. Strategy
  const strategyInputs: StrategyInputs = {
    reporterCount: args.input.reporterCount,
    velocityZscore: 0, // computed by tool; pre-pipeline cheap signal stub
    userRiskTier,
    ruleMatchScore: ruleMatch.score,
    personality: profile.personality,
    tierOverride: profile.tierOverride,
    coldStart,
    threadEscalated,
  };
  const decision = selectStrategy(strategyInputs);

  // 2. Orchestrator
  const context: ToolContext = {
    subredditId: args.input.subredditId,
    correlationId: args.input.correlationId,
    targetKind: args.input.target.kind,
    targetId: args.input.target.id,
    targetBody: args.input.target.body,
    targetAuthorId: args.input.target.author,
    reporterCount: args.input.reporterCount,
    ruleMatchScore: ruleMatch.score,
    threadId: args.input.threadId ?? '',
    threadExcerpts: args.input.threadExcerpts ?? [],
  };
  const orchResult = await new Orchestrator(registry).run({ decision, context });
  const acc = orchResult.accumulator;

  // 2b. Append rule-match as a synthetic evidence row when the score is
  //     meaningful, so the Reasoner can cite [ev-N] for the content match
  //     instead of inventing a slot. signal: 'high' lets it influence
  //     evidence convergence + earlier stop conditions.
  if (ruleMatch.score >= 0.2 && ruleMatch.matchedRule) {
    acc.append({
      tool: 'policy_match',
      status: 'success',
      summary: `Content matches "${ruleMatch.matchedRule.slice(0, 80)}" (${Math.round(
        ruleMatch.score * 100,
      )}% term overlap)`,
      latencyMs: 1,
      detail: {
        matched_rule: ruleMatch.matchedRule,
        matched_terms: ruleMatch.matchedTerms,
        matches: [{ similarity: ruleMatch.score }],
        signal: ruleMatch.score >= 0.5 ? 'high' : 'normal',
      },
    });
  }

  // 3. Reasoner (with validate + retry)
  const isPartial =
    orchResult.earlyStopped && orchResult.stopReason !== 'converged';
  const preset = getPreset(profile.personality);
  const messages = buildMessages({
    accumulator: acc,
    personality: profile.personality,
    personalityPhrasing: preset.promptPhrasing,
    region: profile.region,
    rules: profile.rules,
    targetKind: args.input.target.kind,
    targetId: args.input.target.id,
    targetTitle: args.input.target.title,
    targetBody: args.input.target.body,
    targetAuthor: args.input.target.author,
    reporterCount: args.input.reporterCount,
    tier: decision.tier,
    toolsRun: orchResult.toolsRun,
    isPartial,
    coldStart,
  });

  let reasonerResult: ReasonerResult | null = null;
  let validationFlag = false;

  if (decision.reasonerRequired) {
    reasonerResult = await reasonWithRetry({
      reasoner: new Reasoner(llm),
      messages,
      accumulator: acc,
      correlationId: args.input.correlationId,
    });
    validationFlag = reasonerResult === null;
  }

  const reasonerOutput: ReasonerOutput = reasonerResult?.output ?? fallbackOutput();

  // 4. Calibrate
  const ruleMatchStrength = extractRuleMatchStrength(acc);
  const signals = extractEvidenceSignals(acc);
  const cal = calibrate({
    llmSelfReport: reasonerOutput.raw_confidence,
    evidenceConvergence: computeEvidenceConvergence(signals),
    subredditAccuracy: 0.5,
    ruleMatchStrength,
    validationPassed: !validationFlag,
    coldStart,
    isPartial,
    recommendation: reasonerOutput.recommendation,
  });

  // 4b. Deterministic explainability surfaces (Features 1, 2, 4, 5, 6, 7, 8).
  //     Read existing signals + alignment snapshot in parallel — no extra LLM.
  const priorRemovalsFromInv = extractPriorRemovalsCount(acc);
  const escalation = deriveEscalation(acc);

  const authorSignal = deriveAuthorSignal({
    priorViolations: userMem?.priorViolations ?? 0,
    priorApprovals: userMem?.priorApprovals ?? 0,
    hasHistory: userMem !== null,
    priorRemovalsFromInvestigations: priorRemovalsFromInv,
  });

  const policyEvidenceId =
    acc.successfulEntries().find((e) => e.tool === 'policy_match')?.id ?? null;
  const ruleMatches = buildRuleMatchDisplay({
    match: ruleMatch,
    ruleMatchEvidenceId: policyEvidenceId,
  });

  const reasonerForKey = reasonerResult?.output ?? reasonerOutput;
  const confidenceFactors = deriveConfidenceFactors({
    calibratedConfidence: cal.calibratedConfidence,
    breakdown: {
      llmSelfReport: cal.llmSelfReport,
      evidenceConvergence: cal.evidenceConvergence,
      subredditAccuracy: cal.subredditAccuracy,
      ruleMatchStrength: cal.ruleMatchStrength,
    },
    validationPassed: !validationFlag,
    isPartial,
    coldStart,
    ruleMatchScore: ruleMatch.score,
    escalationLevel: escalation.level,
    authorSignal,
    recommendation: reasonerOutput.recommendation,
  });

  const keyFactors = deriveKeyFactors({
    ruleMatchScore: ruleMatch.score,
    authorSignal,
    velocityZscore: 0, // velocity z is computed inside tools; surface via evidence convergence instead
    reporterCount: args.input.reporterCount,
    escalationLevel: escalation.level,
    evidenceConvergence: cal.evidenceConvergence,
    priorRemovals: priorRemovalsFromInv,
    recommendation: reasonerForKey.recommendation,
  });

  const priority = computePriority({
    calibratedConfidence: cal.calibratedConfidence,
    reporterCount: args.input.reporterCount,
    velocityZscore: extractVelocityZscore(acc),
    userRiskTier: (userMem?.riskTier ?? userRiskTier) as 'new' | 'trusted' | 'neutral' | 'watched',
    priorRemovals: priorRemovalsFromInv,
    escalationLevel: escalation.level,
    ruleMatchScore: ruleMatch.score,
    recommendation: reasonerForKey.recommendation,
  });

  const alignmentSnapshot = await readAlignmentSnapshot(args.input.subredditId).catch(() => ({
    rate: null as number | null,
    total: 0,
    aligned: 0,
  }));

  // 5. Assemble verdict
  const totalMs = Date.now() - t0;
  const verdict: Verdict = {
    correlationId: args.input.correlationId,
    tier: decision.tier,
    riskTier: reasonerOutput.risk_tier,
    recommendation: reasonerOutput.recommendation,
    calibratedConfidence: cal.calibratedConfidence,
    rationale: reasonerOutput.rationale,
    topEvidence: reasonerOutput.top_evidence_ids
      .map((id) => acc.byId(id))
      .filter((e): e is NonNullable<typeof e> => e !== undefined)
      .slice(0, 3)
      .map((e) => ({ id: e.id, summary: e.summary, tool: e.tool })),
    timeline: acc.entries().map((e) => ({
      tool: e.tool,
      verb: TOOL_VERBS[e.tool as ToolName] ?? `Ran ${e.tool}`,
      status: e.status,
      latencyMs: e.latencyMs,
      evidenceIds: [e.id],
    })),
    confidenceBreakdown: {
      llmSelfReport: cal.llmSelfReport,
      evidenceConvergence: cal.evidenceConvergence,
      subredditAccuracy: cal.subredditAccuracy,
      ruleMatchStrength: cal.ruleMatchStrength,
    },
    modelReasoner: reasonerResult?.model ?? '',
    modelSummarizer: '',
    costUsd: round6(reasonerResult?.costUsd ?? 0),
    latencyMs: totalMs,
    validationFlag,
    degraded: reasonerResult === null,
    coldStart,
    priority: {
      score: priority.score,
      bucket: priority.bucket,
      headline: priorityHeadline(priority.bucket),
      drivers: priority.drivers,
    },
    authorSignal,
    escalation,
    confidenceFactors,
    keyFactors,
    ruleMatches,
    alignment: {
      rate: alignmentSnapshot.rate,
      sampleSize: alignmentSnapshot.total,
      aligned: alignmentSnapshot.aligned,
    },
    contentFindings: Array.isArray(reasonerOutput.content_findings)
      ? reasonerOutput.content_findings.slice(0, 6)
      : [],
  };

  // 6. Persist (best-effort; investigation row used by prior_actions next time).
  try {
    await persistInvestigation({
      subId: args.input.subredditId,
      authorId: args.input.target.author,
      targetId: args.input.target.id,
      targetKind: args.input.target.kind,
      targetBody: args.input.target.body,
      verdict,
    });
  } catch (e) {
    console.warn('engine.persist_failed', {
      correlation_id: verdict.correlationId,
      err: e instanceof Error ? e.message : String(e),
    });
  }

  // 7. Bump subreddit-level stats counters. Best-effort.
  try {
    await bumpInvestigationStats(args.input.subredditId, verdict);
  } catch (e) {
    console.warn('engine.stats_bump_failed', {
      correlation_id: verdict.correlationId,
      err: e instanceof Error ? e.message : String(e),
    });
  }

  return verdict;
}

async function reasonWithRetry(args: {
  reasoner: Reasoner;
  messages: Parameters<Reasoner['reason']>[0]['messages'];
  accumulator: EvidenceAccumulator;
  correlationId: string;
}): Promise<ReasonerResult | null> {
  let first: ReasonerResult | null = null;
  try {
    first = await args.reasoner.reason({
      messages: args.messages,
      correlationId: args.correlationId,
    });
  } catch (e) {
    console.warn('reasoner.failed', { attempt: 1, err: errMsg(e) });
    return null;
  }

  const v1 = validateCitations(
    first.output.rationale,
    args.accumulator,
    first.output.cited_evidence_ids,
  );
  if (v1.passed) return first;
  console.warn('reasoner.validation_failed', { attempt: 1, reason: v1.reason });

  const corrective = buildCorrectiveMessages({
    priorMessages: args.messages,
    priorResponse: first.rawText,
    reason: v1.reason,
    details: JSON.stringify(v1.details),
  });

  try {
    const second = await args.reasoner.reason({
      messages: corrective,
      correlationId: `${args.correlationId}:retry`,
    });
    const v2 = validateCitations(
      second.output.rationale,
      args.accumulator,
      second.output.cited_evidence_ids,
    );
    if (v2.passed) return second;
    console.warn('reasoner.validation_failed', { attempt: 2, reason: v2.reason });
    return null;
  } catch (e) {
    console.warn('reasoner.failed', { attempt: 2, err: errMsg(e) });
    return null;
  }
}

function fallbackOutput(): ReasonerOutput {
  return {
    risk_tier: 'LOW' as RiskTier,
    recommendation: 'NO_RECOMMENDATION' as Recommendation,
    rationale:
      'ModPilot was unable to produce a recommendation for this report. The evidence has been collected and is available for review [ev-1].',
    top_evidence_ids: ['ev-1'],
    raw_confidence: 0,
    cited_evidence_ids: ['ev-1'],
    flags: ['reasoner_failed'],
    content_findings: [],
  };
}

function extractPriorRemovalsCount(acc: EvidenceAccumulator): number {
  const entry = acc.successfulEntries().find((e) => e.tool === 'prior_actions');
  if (!entry) return 0;
  const removals = entry.detail.removals;
  return typeof removals === 'number' ? removals : 0;
}

function extractVelocityZscore(acc: EvidenceAccumulator): number {
  const entry = acc.successfulEntries().find((e) => e.tool === 'report_velocity');
  if (!entry) return 0;
  const z = entry.detail.z_score;
  return typeof z === 'number' ? z : 0;
}

function extractRuleMatchStrength(acc: EvidenceAccumulator): number {
  for (const e of acc.successfulEntries()) {
    if (e.tool === 'policy_match') {
      const matches = e.detail.matches;
      if (Array.isArray(matches) && matches.length > 0) {
        const first = matches[0];
        if (first && typeof first === 'object' && 'similarity' in first) {
          const sim = (first as { similarity?: unknown }).similarity;
          return typeof sim === 'number' ? sim : 0;
        }
      }
    }
  }
  return 0;
}

function extractEvidenceSignals(acc: EvidenceAccumulator): number[] {
  const out: number[] = [];
  for (const e of acc.successfulEntries()) {
    if (e.tool === 'policy_match') {
      const m = e.detail.matches;
      if (Array.isArray(m) && m.length > 0 && m[0] && typeof m[0] === 'object') {
        const sim = (m[0] as { similarity?: unknown }).similarity;
        out.push(typeof sim === 'number' ? sim : 0);
      } else {
        out.push(0);
      }
    } else if (e.tool === 'report_velocity') {
      const z = e.detail.z_score;
      const zn = typeof z === 'number' ? z : 0;
      out.push(Math.min(Math.abs(zn) / 5, 1));
    } else {
      out.push(0.5);
    }
  }
  return out;
}

function round6(n: number): number {
  return Math.round(n * 1_000_000) / 1_000_000;
}

function errMsg(e: unknown): string {
  return e instanceof Error ? e.message : String(e);
}
